### PR TITLE
feat(study-search): add a studies counting endpoint

### PR DESCRIPTION
**Pull request by mabw-rte**
Context:
Related to ANT-1107 (tags-db) and ANT-1106 (permissions-db), it happens that front-end can not predict the total number of studies matching some given filtering parameters, to perform the pagination properly.

Solution:
Add an endpoint that return the total studies count.

### CODE REVIEW

File: antarest/study/web/studies_blueprint.py [hunk 1]
```diff
@@ -168,6 +157,77 @@ def get_studies(
...
+        """
+
+        logger.info("Counting matching studies", extra={"user": current_user.id})
+        params = RequestParameters(user=current_user)
+
+        user_list = [int(v) for v in _split_comma_separated_values(users)]
+
+        if not params.user:
+            logger.error("FAIL permission: user is not logged")
+            raise UserHasNotPermissionError()
```
Comment (laurent-laporte-pro) [inline antarest/study/web/studies_blueprint.py]: Le message d'erreur pourrait être dans l'exception au lieu du log.

File: antarest/study/web/studies_blueprint.py [hunk 2]
```diff
@@ -168,6 +157,77 @@ def get_studies(
...
+
+        total_matching_studies = study_service.count_studies(
```
Comment (laurent-laporte-pro) [inline antarest/study/web/studies_blueprint.py]: Tu aimes bien les noms de variable de 22 caractères ? On peut pas mettre simplement `count` ?

**Review by laurent-laporte-pro — DISMISSED** [review]
J'ai fais des commentaires dans la PR #1931, que tu peux valider.
Fais ensuite des corrections dans cette PR.

File: antarest/study/web/studies_blueprint.py [hunk 3]
```diff
@@ -171,6 +155,76 @@ def get_studies(
...
+        - `groups`: Comma-separated group IDs for studies to be selected.
+        - `tags`: Comma-separated tags for studies to be selected.
+        - `studyIds`: Comma-separated IDs of studies to be selected.
+        - `exists`: Whether to limit the selection based on studies' existence on disk.
+        - `workspace`: to limit studies selection based on their workspace.
+        - `folder`: to limit studies selection based on their folder.
+
+        Returns:
+        - An integer representing the total number of studies matching the filters above and the user permissions.
+        """
```
Comment (hdinia) [inline antarest/study/web/studies_blueprint.py:198]: it's a bit redundant with the `description` option of pydantic, that also document the swagger

Comment (laurent-laporte-pro) [reply antarest/study/web/studies_blueprint.py:198]: Historiquement, nous avions adopté une approche consistant à documenter chaque paramètre dans la docstring des endpoints.
Cependant, ayant découvert que la description des paramètres au niveau des `Query` est correctement intégrée dans la doc du Swagger, je suis convaincu que cette pratique est désormais redondante et peut être abandonnée.

**Review by laurent-laporte-pro — DISMISSED** [review]
La docstring du endpoint `count_studies` (et aussi `get_studies`) peut être simplifiée car il se trouve que les paramètres `Query` sont déjà bien documentés. Tu peux simplifier si tu le désire.

File: antarest/study/web/studies_blueprint.py (hunk 3 above)
Comment (mabw-rte) [reply antarest/study/web/studies_blueprint.py:198]: Cela sera traite dans une PR de type refactor(bp-docs).

File: antarest/study/web/studies_blueprint.py [hunk 4]
```diff
@@ -102,23 +92,17 @@ def get_studies(
...
+            0, description="Number of studies per page (0 = no limit).", alias="pageSize"
         ),
     ) -> t.Dict[str, StudyMetadataDTO]:
         """
         Get the list of studies matching the specified criteria.
 
         Args:
+
         - `name`: Filter studies based on their name. Case-insensitive search for studies
-          whose name contains the specified value.
```
Comment (MartinBelthle) [inline antarest/study/web/studies_blueprint.py:121]: Why did you remove this ?

File: antarest/study/web/studies_blueprint.py [hunk 5]
```diff
@@ -171,6 +155,76 @@ def get_studies(
...
+        tags=[APITag.study_management],
+        summary="Count Studies",
+    )
+    def count_studies(
+        current_user: JWTUser = Depends(auth.get_current_user),
+        name: str = Query("", description="Case-insensitive: filter studies based on their name.", alias="name"),
+        managed: t.Optional[bool] = Query(None, description="Management status filter."),
+        archived: t.Optional[bool] = Query(None, description="Archive status filter."),
+        variant: t.Optional[bool] = Query(None, description="Variant status filter."),
+        versions: str = Query("", description="Comma-separated versions filter.", regex=QUERY_REGEX),
```
Comment (MartinBelthle) [inline antarest/study/web/studies_blueprint.py:169]: Why do we use str instead of list if we have to split it afterwards ?

File: tests/integration/studies_blueprint/test_get_studies.py [hunk 6]
```diff
@@ -454,6 +454,16 @@ def test_study_listing(
         study_map = res.json()
         assert not all_studies.intersection(study_map)
         assert all(map(lambda x: pm(x) in [PublicMode.READ, PublicMode.FULL], study_map.values()))
+        # test pagination
+        res = client.get(
+            STUDIES_URL,
+            headers={"Authorization": f"Bearer {john_doe_access_token}"},
+            params={"pageNb": 1, "pageSize": 2},
+        )
+        if len(study_map) > 2:
```
Comment (MartinBelthle) [inline tests/integration/studies_blueprint/test_get_studies.py]: I think you should remove this from your test as it's a bit confusing :)
Same goes for every other occurence

File: tests/integration/studies_blueprint/test_get_studies.py [hunk 7]
```diff
@@ -1343,12 +1374,24 @@ def test_get_studies__access_permissions(self, client: TestClient, admin_access_
             study_map = res.json()
             assert not expected_studies.difference(set(study_map))
             assert not all_studies.difference(expected_studies).intersection(set(study_map))
+            # test pagination
+            if len(expected_studies) > 2:
```
Comment (MartinBelthle) [inline tests/integration/studies_blueprint/test_get_studies.py]: Same here

File: tests/integration/studies_blueprint/test_get_studies.py [hunk 8]
```diff
@@ -1473,3 +1516,21 @@ def test_get_studies__invalid_parameters(
         assert res.status_code == INVALID_PARAMS_STATUS_CODE, res.json()
         description = res.json()["description"]
         assert re.search(r"could not be parsed to a boolean", description), f"{description=}"
+
+
+def test_studies_counting(client: TestClient, admin_access_token: str, user_access_token: str) -> None:
```
Comment (MartinBelthle) [inline tests/integration/studies_blueprint/test_get_studies.py:1533]: Put this in the class of the test as the ones before

File: tests/integration/studies_blueprint/test_get_studies.py [hunk 9]
```diff
@@ -1473,3 +1516,21 @@ def test_get_studies__invalid_parameters(
...
+    # test admin studies count
+    res = client.get(STUDIES_URL, headers={"Authorization": f"Bearer {admin_access_token}"})
+    assert res.status_code == 200, res.json()
+    expected_studies_count = len(res.json())
+    res = client.get(STUDIES_URL + "/count", headers={"Authorization": f"Bearer {admin_access_token}"})
+    assert res.status_code == 200, res.json()
+    assert res.json() == expected_studies_count
+
+    # test user studies count
```
Comment (MartinBelthle) [inline tests/integration/studies_blueprint/test_get_studies.py]: Do a loop around `admin_access_token` and `user_access_token` as it's exactly the same code

File: tests/study/test_repository.py [hunk 10]
```diff
@@ -66,24 +66,31 @@ def test_get_all__general_case(
...
         _ = [s.additional_data for s in all_studies]
         _ = [s.tags for s in all_studies]
     assert len(db_recorder.sql_statements) == 1, str(db_recorder)
 
+    # test that the expected studies are returned
     if expected_ids is not None:
         assert {s.id for s in all_studies} == expected_ids
 
+    # test pagination
+    if len(expected_ids) > 2:
```
Comment (MartinBelthle) [inline tests/study/test_repository.py]: Perhaps remove this `len` check for the same reason I explained above

File: tests/study/test_repository.py [hunk 11]
```diff
@@ -1004,3 +1177,71 @@ def test_update_tags(
     # Check that only "Tag1" and "Tag3" are present in the database
     tags = db_session.query(Tag).all()
     assert {tag.label for tag in tags} == {"Tag1", "Tag3"}
+
+
+@pytest.mark.parametrize(
```
Comment (MartinBelthle) [inline tests/study/test_repository.py:1131]: I really like the diversification here and i think you could do these a bit more for the other tests you added. For instance, i think you always check the pagination with these parameters `page_nb=1, page_size=2` and it could be different

File: tests/study/test_repository.py [hunk 12]
```diff
@@ -1004,3 +1177,71 @@ def test_update_tags(
...
+    "managed, study_ids, exists, expected_ids",
+    [
+        (None, [], False, {"5", "6"}),
```
Comment (MartinBelthle) [inline tests/study/test_repository.py:1134]: But it seems that this duplicates the first test of the file :/ Or i'm missing something

**Review by MartinBelthle — CHANGES_REQUESTED** [review]

File: tests/study/test_repository.py (hunk 12 above)
Comment (mabw-rte) [reply tests/study/test_repository.py:1134]: @MartinBelthle  
I could not put this test in the listing class as the purpose is not the same. 
Counting returns an `int` meanwhile listing returns a list of `Studies`.
Regarding the similarity with previous part is partially due to the fact that @laurent-laporte-pro  asked only for general case.

File: tests/study/test_repository.py (hunk 11 above)
Comment (mabw-rte) [reply tests/study/test_repository.py:1131]: pagination is not relevant to the counting endpoint

File: antarest/study/web/studies_blueprint.py (hunk 5 above)
Comment (mabw-rte) [reply antarest/study/web/studies_blueprint.py:169]: It is easier for the front, or at least this is what I was told.

File: antarest/study/web/studies_blueprint.py (hunk 4 above)
Comment (mabw-rte) [reply antarest/study/web/studies_blueprint.py:121]: I did not remove it, It is the black reformating. Indeed, I replaced the regex by QUERY_REGEX, as it is used many times.

File: tests/integration/studies_blueprint/test_get_studies.py (hunk 6 above)
Comment (mabw-rte) [reply tests/integration/studies_blueprint/test_get_studies.py]: yes, I made a generalization for the testing.

**Review by MartinBelthle — APPROVED** [review]
